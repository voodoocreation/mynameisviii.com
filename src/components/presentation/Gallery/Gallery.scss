@import "../../../scss/variables", "../../../scss/mixins";

.Gallery {
  &--description {
    @include panel-1();

    margin-bottom: $spacing;
    padding: $spacing-half $spacing;
  }

  &--images {
    padding-bottom: $spacing;
  }

  &--image {
    flex-basis: 50%;
    padding: $spacing-half;
    cursor: pointer;
  }

  .ImageGallery {
    display: flex;
    flex-flow: row wrap;
    margin: -($spacing-half);
  }

  @include when-above($screen-xs-min) {
    &--image {
      flex-basis: 33.333333%;
    }
  }

  @include when-above($screen-sm-min) {
    &-image {
      flex-basis: 25%;
    }
  }

  @include when-above($screen-md-min) {
    &--image {
      flex-basis: 20%;
    }
  }

  @include when-above($screen-lg-min) {
    &--image {
      flex-basis: 16.666666%;
    }
  }
}
