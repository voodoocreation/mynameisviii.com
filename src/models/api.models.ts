import { arrayToAssoc } from "../transformers/transformData";

interface IWithSlug {
  slug: string;
}

export type TLastEvaluatedKey<K extends string> = Record<
  K | "isActive" | "slug",
  string
>;

export interface IRawDynamoResponse<T extends {}, K extends string> {
  items: T[];
  lastEvaluatedKey?: TLastEvaluatedKey<K>;
}

export interface IDynamoResponse<T extends IWithSlug, K extends string> {
  readonly items: Record<string, T>;
  readonly lastEvaluatedKey?: TLastEvaluatedKey<K>;
}

export interface IRawS3Response<T extends {}> {
  isTruncated?: boolean;
  items: T[];
}

export interface IS3Response<T extends IWithSlug> {
  readonly isTruncated: boolean;
  readonly items: Record<string, T>;
}

export const dynamoResponse = <T extends IWithSlug, K extends string>(
  options: IRawDynamoResponse<T, K>
): IDynamoResponse<T, K> => ({
  items: arrayToAssoc(options.items, "slug"),
  lastEvaluatedKey: options.lastEvaluatedKey
});

export const s3Response = <T extends IWithSlug>(
  options: IRawS3Response<T>
): IS3Response<T> => ({
  isTruncated: !!options.isTruncated,
  items: arrayToAssoc(options.items, "slug")
});
