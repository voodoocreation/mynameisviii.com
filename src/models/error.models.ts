export interface IError {
  readonly message: string;
  readonly status: number;
}

export const error = (options: Partial<IError> = {}): IError => ({
  message: options.message || "An error has occurred",
  status: options.status || 500,
});
