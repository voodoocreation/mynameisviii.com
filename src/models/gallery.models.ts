import dayjs from "dayjs";

export interface IRawGallery {
  description?: string;
  imageUrl?: string;
  images?: Array<Partial<IGalleryImage>>;
  modifiedAt?: string;
  slug?: string;
  title?: string;
}

export interface IGallery {
  readonly description?: string;
  readonly imageUrl: string;
  readonly images?: IGalleryImage[];
  readonly modifiedAt: string;
  readonly slug: string;
  readonly title: string;
}

export interface IGalleryImage {
  readonly imageUrl: string;
  readonly modifiedAt: string;
}

export const galleryImage = (
  options: Partial<IGalleryImage> = {}
): IGalleryImage => ({
  imageUrl: options.imageUrl || "",
  modifiedAt: options.modifiedAt || dayjs().toISOString()
});

export const gallery = (options: IRawGallery = {}): IGallery => ({
  description: options.description,
  imageUrl: options.imageUrl || "",
  images: options.images ? options.images.map(galleryImage) : undefined,
  modifiedAt: options.modifiedAt || dayjs().toISOString(),
  slug: options.slug || "",
  title: options.title || ""
});
