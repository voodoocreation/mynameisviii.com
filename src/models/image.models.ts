export interface IImage {
  readonly imageUrl: string;
  readonly title: string;
}

export const image = (options: Partial<IImage> = {}): IImage => ({
  imageUrl: options.imageUrl || "",
  title: options.title || ""
});
