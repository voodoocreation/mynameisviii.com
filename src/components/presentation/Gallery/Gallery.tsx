import * as React from "react";

import { IGallery } from "../../../models/root.models";
import Schema from "../../schema/Gallery";
import Image from "../Image/Image";
import ImageGallery from "../ImageGallery/ImageGallery";
import PageHeader from "../PageHeader/PageHeader";

import "./Gallery.scss";

interface IProps extends IGallery {
  onGalleryInteraction?: (type: string, index?: number) => void;
}

export default class Gallery extends React.Component<IProps> {
  public render() {
    const { onGalleryInteraction, ...gallery } = this.props;

    return (
      <article className="Gallery">
        <PageHeader>{gallery.title}</PageHeader>

        <div className="Gallery--description">
          <p>{gallery.description}</p>
        </div>

        {gallery.images && gallery.images.length > 0 ? (
          <section className="Gallery--images">
            <ImageGallery
              onItemClick={this.onGalleryInteraction("itemClick")}
              onNext={this.onGalleryInteraction("next")}
              onPrevious={this.onGalleryInteraction("previous")}
              onModalClose={this.onGalleryInteraction("modalClose")}
            >
              {gallery.images.map((image, index) => (
                <Image
                  alt={`${gallery.title} - ${index + 1}`}
                  className="Gallery--image"
                  key={image.imageUrl}
                  src={image.imageUrl}
                />
              ))}
            </ImageGallery>
          </section>
        ) : null}

        <Schema {...gallery} />
      </article>
    );
  }

  private onGalleryInteraction = (type: string) => (index?: number) => {
    if (this.props.onGalleryInteraction) {
      this.props.onGalleryInteraction(type, index);
    }
  };
}
