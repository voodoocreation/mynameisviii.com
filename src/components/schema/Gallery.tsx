import * as React from "react";

import { absoluteUrl } from "../../helpers/dataTransformers";
import { IGallery } from "../../models/root.models";
import Schema from "./Schema";

const Gallery: React.FC<IGallery> = props => (
  <Schema
    {...{
      "@id": absoluteUrl(`/galleries/${props.slug}`),
      "@type": "ImageGallery",
      associatedMedia:
        props.images && props.images.length > 0
          ? props.images.map(image => ({
              "@type": "ImageObject",
              contentUrl: image.imageUrl,
              dateModified: image.modifiedAt
            }))
          : undefined,
      dateModified: props.modifiedAt,
      description: props.description,
      name: props.title,
      primaryImageOfPage: props.imageUrl,
      thumbnailUrl: props.imageUrl
    }}
  />
);

export default Gallery;
