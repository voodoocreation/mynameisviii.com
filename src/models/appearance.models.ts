import dayjs from "dayjs";

import { BOOLEAN } from "../constants/api.constants";
import { STATUS, TYPE } from "../constants/appearance.constants";
import { IImage, image } from "./image.models";
import { ILocation, location } from "./location.models";
import { IOffer, offer } from "./offer.models";
import { IOrganization, organization } from "./organization.models";
import { IPerformer, IRawPerformer, performer } from "./performer.models";

export interface IRawAppearance {
  acts?: Array<Partial<IRawPerformer>>;
  audience?: string;
  description?: string;
  finishingAt?: string;
  imageUrl?: string;
  images?: IImage[];
  isActive?: BOOLEAN;
  location?: Partial<ILocation>;
  ogImageUrl?: string;
  organizer?: Partial<IOrganization>;
  rsvpUrl?: string;
  sales?: Array<Partial<IOffer>>;
  slug?: string;
  startingAt?: string;
  status?: STATUS;
  title?: string;
  type?: TYPE;
}

export interface IAppearance {
  readonly acts: IPerformer[];
  readonly audience?: string;
  readonly description: string;
  readonly finishingAt: string;
  readonly imageUrl: string;
  readonly images: IImage[];
  readonly isActive: boolean;
  readonly location: ILocation;
  readonly ogImageUrl: string;
  readonly organizer: IOrganization;
  readonly rsvpUrl?: string;
  readonly sales: IOffer[];
  readonly slug: string;
  readonly startingAt: string;
  readonly status: STATUS;
  readonly title: string;
  readonly type: TYPE;
}

export const appearance = (options: IRawAppearance = {}): IAppearance => ({
  acts: options.acts ? options.acts.map(performer) : [],
  audience: options.audience,
  description: options.description || "",
  finishingAt: options.finishingAt || dayjs().toISOString(),
  imageUrl: options.imageUrl || "",
  images: options.images ? options.images.map(image) : [],
  isActive: (options.isActive && options.isActive === BOOLEAN.TRUE) || false,
  location: location(options.location),
  ogImageUrl: options.ogImageUrl || "",
  organizer: organization(options.organizer),
  rsvpUrl: options.rsvpUrl,
  sales: options.sales ? options.sales.map(offer) : [],
  slug: options.slug || "",
  startingAt: options.startingAt || dayjs().toISOString(),
  status: options.status || STATUS.SCHEDULED,
  title: options.title || "",
  type: options.type || TYPE.MUSIC,
});
