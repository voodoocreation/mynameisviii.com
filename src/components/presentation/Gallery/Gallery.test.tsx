import { gallery, galleryImage } from "../../../models/root.models";
import WrapperWithIntl from "../../../utilities/WrapperWithIntl";
import Gallery from "./Gallery";

const component = new WrapperWithIntl(Gallery).withDefaultProps(
  gallery({
    description: "Description",
    imageUrl: "Image URL",
    modifiedAt: "2017-10-10T18:00:00",
    slug: "test-1",
    title: "Title"
  })
);

const images = [
  galleryImage({ imageUrl: "Image 1" }),
  galleryImage({ imageUrl: "Image 2" }),
  galleryImage({ imageUrl: "Image 3" })
];

describe("[presentation] <Gallery />", () => {
  describe("when there are images defined", () => {
    const onGalleryInteraction = jest.fn();
    const wrapper = component
      .withProps({
        images,
        onGalleryInteraction
      })
      .mount();

    it("renders all images", () => {
      expect(wrapper.find("Image.Gallery--image")).toHaveLength(3);
    });

    it("clicks an image", () => {
      wrapper
        .find("Image.Gallery--image")
        .first()
        .simulate("click");
    });

    it("calls onGalleryInteraction prop", () => {
      expect(onGalleryInteraction).toHaveBeenCalledTimes(1);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe("when there are no images", () => {
    const wrapper = component
      .withProps({
        images: undefined
      })
      .render();

    it("doesn't render the images container", () => {
      expect(wrapper.find(".Gallery--images")).toHaveLength(0);
    });

    it("matches snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("when onGalleryInteraction isn't defined", () => {
    const wrapper = component
      .withProps({
        images
      })
      .mount();

    it("doesn't throw an error when clicking a gallery item", () => {
      expect(() => {
        wrapper
          .find("Image.Gallery--image")
          .first()
          .simulate("click");
      }).not.toThrowError();
    });
  });
});
