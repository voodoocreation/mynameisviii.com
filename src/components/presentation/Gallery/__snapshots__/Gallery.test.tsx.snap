// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <Gallery /> when there are images defined matches snapshot 1`] = `
<article
  class="Gallery"
>
  <header
    class="PageHeader"
  >
    <h1>
      Title
    </h1>
  </header>
  <div
    class="Gallery--description"
  >
    <p>
      Description
    </p>
  </div>
  <section
    class="Gallery--images"
  >
    <div
      class="ImageGallery"
    >
      <figure
        class="Image Gallery--image isLoading"
        role="button"
        tabindex="0"
      >
        <div
          class="Image--wrapper"
        >
          <img
            alt="Title - 1"
            src="Image 1"
          />
          <span
            class="Loader"
          >
            <div
              class="Loader--animation"
            >
              <span />
              <span />
              <span />
            </div>
          </span>
        </div>
      </figure>
      <figure
        class="Image Gallery--image isLoading"
        role="button"
        tabindex="0"
      >
        <div
          class="Image--wrapper"
        >
          <img
            alt="Title - 2"
            src="Image 2"
          />
          <span
            class="Loader"
          >
            <div
              class="Loader--animation"
            >
              <span />
              <span />
              <span />
            </div>
          </span>
        </div>
      </figure>
      <figure
        class="Image Gallery--image isLoading"
        role="button"
        tabindex="0"
      >
        <div
          class="Image--wrapper"
        >
          <img
            alt="Title - 3"
            src="Image 3"
          />
          <span
            class="Loader"
          >
            <div
              class="Loader--animation"
            >
              <span />
              <span />
              <span />
            </div>
          </span>
        </div>
      </figure>
    </div>
  </section>
  <script
    type="application/ld+json"
  >
    {"@id":"http://localhost:80/galleries/test-1","@type":"ImageGallery","associatedMedia":[{"@type":"ImageObject","contentUrl":"Image 1","dateModified":"2017-12-31T11:00:00.000Z"},{"@type":"ImageObject","contentUrl":"Image 2","dateModified":"2017-12-31T11:00:00.000Z"},{"@type":"ImageObject","contentUrl":"Image 3","dateModified":"2017-12-31T11:00:00.000Z"}],"dateModified":"2017-10-10T18:00:00","description":"Description","name":"Title","primaryImageOfPage":"Image URL","thumbnailUrl":"Image URL","@context":"http://schema.org/"}
  </script>
</article>
`;

exports[`[presentation] <Gallery /> when there are no images matches snapshot 1`] = `
<article
  class="Gallery"
>
  <header
    class="PageHeader"
  >
    <h1>
      Title
    </h1>
  </header>
  <div
    class="Gallery--description"
  >
    <p>
      Description
    </p>
  </div>
  <script
    type="application/ld+json"
  >
    {"@id":"http://localhost:80/galleries/test-1","@type":"ImageGallery","dateModified":"2017-10-10T18:00:00","description":"Description","name":"Title","primaryImageOfPage":"Image URL","thumbnailUrl":"Image URL","@context":"http://schema.org/"}
  </script>
</article>
`;
